### PR TITLE
Automatic swap refund bot

### PR DESCRIPTION
**Pull request by denalimarsh**
Introduces RefundBot, a swap refund process that scans Kava and Binance Chain for refundable swaps and automatically processes them. Works on testnet. ~~Once approved, I'll launch a server for RefundBot and put it live on mainnet.~~ **Edit:** RefundBot is live on mainnet.

Example testnet refund cron routine output:
```bash
[2020-06-15T08:51:32.754Z] Kava refundable swap count: 1
[2020-06-15T08:51:32.757Z]      Refunding swap 251165847a4f65efe47174339b76cf65f817e5fd700018d51a8dad89e4b2ec1f
[2020-06-15T08:51:32.790Z]      Tx hash: D3F38B67C2E0BF073E1BC1111CAEF09F81141B1DA568C2C462AE3E909B4F2D6E
[2020-06-15T08:51:39.458Z] Binance Chain refundable swap count: 8
[2020-06-15T08:51:39.458Z]      Refunding swap 5a044d1e9fe39f1065a61bfa2dc66b3ead80ddf84b960f8fbbfc7d39609bad10
[2020-06-15T08:51:40.789Z]      Tx hash: 541F78A34A4BDCE69F8ECE61E362681C47D17746AC43C2F5D3ED7F6602615FBA
[2020-06-15T08:51:43.793Z]      Refunding swap 130cd8ec49960989343f7988c7d728959ad8e8fbca363350d32bb7ff418d1252
[2020-06-15T08:51:45.239Z]      Tx hash: 4EB31E241535EEAA3095CB96EEC7446CB0E28441A570856078ADB071F015FCEC
[2020-06-15T08:51:48.243Z]      Refunding swap eb69463cc2063d230ec9079b91495c309c68ccdf7d8811db1aa488ee72a64dbd
[2020-06-15T08:51:48.879Z]      Tx hash: 68982CEC5F3EC870CC9018CA99C89E289E6854C135573850082C70474128361D
...
```

The RefundBot waits 3 seconds between posting refund txs to Binance Chain, otherwise, an account sequence error is thrown. I've created https://github.com/Kava-Labs/kava-tools/issues/14 for upstreaming this to the binance-chain/javascript-sdk.

Every hour the query offsets are printed. Binance Chain uses a different swap querying strategy and this monitoring allows for debugging and future optimization: 
```bash
[2020-06-15T08:54:12.153Z] Current Binance Chain offsets:
[2020-06-15T08:54:12.154Z]      Offset incoming: 0
[2020-06-15T08:54:12.154Z]      Offset outgoing: 0
```

### CODE REVIEW

**Comment by denalimarsh** [comment]
**NOTE**: This PR updates `@kava-labs/javascript-sdk` to `^2.0.0-beta.6`, which includes significant client changes. Before this PR is merged all other kava-tools should be run on this branch to confirm everything works fine.

File: refund/refund.js [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+        this.offsetOutgoing = offsetOutgoing;
+    }
+
+  /**
+   * Initialize the Kava client
+   * @param {String} lcdURL api endpoint for Kava's rest-server
+   * @param {String} mnemonic Kava address mnemonic
+   * @return {Promise}
+   */
+    async initKavaClient(lcdURL, mnemonic) {
```
Comment (nddeluca) [inline refund/refund.js:28]: Nice init functions 👍 .  If we start using these clients in more bots it'll be easy to extract these into their own functions and inject an already initialized client into bots to share dependencies and make mocking a bit easier down the line.

File: refund/refund.js [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        // Fetch account data so we can manually manage sequence when posting
+        let accountData
+        try {
+            accountData = await Kava.tx.loadMetaData(this.kavaClient.wallet.address, this.kavaClient.baseURI)
+        } catch(e) {
+            console.log(e)
+        }
+
+        // Refund each swap
+        for(let i = 0; i < swapIDs.length; i++) {
```
Comment (nddeluca) [inline refund/refund.js:106]: Can switch to a for of loop if you want here.
```javascript
for(const swapID of swapIDs) {
 ...
}
```

File: refund/refund.js [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+                page++;
+            // If no swaps in batch, don't check the next batch
+            } else {  
+                checkNextBatch = false
+            }
+        }
+    
+        // Calculate each swap's ID as it's not stored in the struct (it's on the interface)
+        let swapIDs = []
+        for(let i = 0; i < expiredSwaps.length; i++) {
```
Comment (nddeluca) [inline refund/refund.js:148]: ditto of on the `for...of`

File: refund/refund.js [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    async refundKavaSwaps() {
+        const swapIDs = await this.getRefundableKavaSwaps();
+        console.log(`Kava refundable swap count: ${swapIDs.length}`)
+
+        // Fetch account data so we can manually manage sequence when posting
+        let accountData
+        try {
+            accountData = await Kava.tx.loadMetaData(this.kavaClient.wallet.address, this.kavaClient.baseURI)
+        } catch(e) {
+            console.log(e)
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply refund/refund.js:102]: fixed in https://github.com/Kava-Labs/kava-tools/pull/15/commits/2c9734ae69925fad2bbc85f791a46506316abb9f

File: refund/refund.js [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+            throw new Error("Kava address mnemonic is required");
+        }
+
+        // Initiate and set Kava client
+        this.kavaClient = new Kava.KavaClient(lcdURL);
+        this.kavaClient.setWallet(mnemonic);
+        try {
+            await this.kavaClient.initChain();
+        } catch (e) {
+            console.log("Error: cannot connect to Kava's lcd server")
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply refund/refund.js]: Good idea, updated in https://github.com/Kava-Labs/kava-tools/pull/15/commits/2c9734ae69925fad2bbc85f791a46506316abb9f

File: refund/refund.js (hunk 2 above)
Comment (denalimarsh) [reply refund/refund.js]: https://github.com/Kava-Labs/kava-tools/pull/15/commits/2c9734ae69925fad2bbc85f791a46506316abb9f

File: refund/refund.js (hunk 3 above)
Comment (denalimarsh) [reply refund/refund.js]: https://github.com/Kava-Labs/kava-tools/pull/15/commits/2c9734ae69925fad2bbc85f791a46506316abb9f

**Pull request merged**